### PR TITLE
Add softmax cross entropy loss with label smoothing support.

### PR DESCRIPTION
**Pull request by alpha0422**
@jjsjann123 @mcarilli 
This PR contains softmax cross entropy loss function with label smoothing support, which will reduce memory usage and improve performance.
A brief unittest results:
```
$ python test_label_smoothing.py        
Max atol idx: 246645668, diff: 0.000001, ref: 0.001428, tst: 0.001429 
Max atol idx: 289230632, diff: 0.000001, ref: 0.001148, tst: 0.001149 
Max atol idx: 60159214, diff: 0.000002, ref: 0.001976, tst: 0.001978  
Max atol idx: 87538934, diff: 0.000001, ref: 0.001067, tst: 0.001066  
Max atol idx: 107560465, diff: 0.000001, ref: 0.001181, tst: 0.001180 
Max atol idx: 35533876, diff: 0.000001, ref: 0.001122, tst: 0.001122  
Max atol idx: 17088608, diff: 0.000001, ref: 0.001122, tst: 0.001122  
Max atol idx: 69708634, diff: 0.000001, ref: 0.001080, tst: 0.001081  
Max atol idx: 170873552, diff: 0.000001, ref: 0.001063, tst: 0.001064 
Max atol idx: 274965302, diff: 0.000002, ref: 0.002668, tst: 0.002670 
.                                                                     
Raw time 19.32 s elapsed for 1000 iterations, norm 0.6245             
Opt time 3.44 s elapsed for 1000 iterations, norm 0.6245              
.                                                                     
----------------------------------------------------------------------
Ran 2 tests in 72.630s                                                
                                                                      
OK                                                                    
```

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -128,6 +128,30 @@ def check_cuda_torch_binary_vs_bare_metal(cuda_dir):
...
 
+if "--xentropy" in sys.argv:
+    from torch.utils.cpp_extension import CUDAExtension
+    sys.argv.remove("--xentropy")
+
+    from torch.utils.cpp_extension import BuildExtension
+    cmdclass['build_ext'] = BuildExtension
+
+    if torch.utils.cpp_extension.CUDA_HOME is None:
+        raise RuntimeError("--cuda_ext was requested, but nvcc was not found.  Are you sure your environment has nvcc available?  If you're installing within a container from https://hub.docker.com/r/pytorch/pytorch, only images whose names contain 'devel' will provide nvcc.")
```
Comment (jjsjann123) [inline setup.py]: I'm gonna shamelessly point this out `--cuda_ext` should be renamed.
I should have fixed the same thing up there for `bnp` as well

File: apex/contrib/csrc/xentropy/interface.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+at::Tensor softmax_xentropy_backward(
+    const at::Tensor &grad_loss,
+    const at::Tensor &logits,
+    const at::Tensor &max_log_sum_exp,
+    const at::Tensor &labels,
+    const float smoothing)  {
+    CHECK_INPUT(grad_loss);
+    CHECK_INPUT(max_log_sum_exp);
+    CHECK_INPUT(labels);
+
```
Comment (jjsjann123) [inline apex/contrib/csrc/xentropy/interface.cpp]: Looks missed `CHECK_INPUT(logits)`

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+struct LogSoftMaxForwardEpilogue {
+  __device__ __forceinline__ LogSoftMaxForwardEpilogue(AccumT max_input, AccumT sum)
+    : logsum(max_input + std::log(sum)) {}
+
+  __device__ __forceinline__ LogSoftMaxForwardEpilogue(AccumT max_log_sum_exp)
+    : logsum(max_log_sum_exp) {}
+
+  __device__ __forceinline__ OutT operator()(T input) const {
+    return static_cast<OutT>(input - logsum);
+}
```
Comment (jjsjann123) [inline apex/contrib/csrc/xentropy/xentropy_kernel.cu]: indentation

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+
+
+const int max_threads = 1024;
+
+inline dim3 SoftMax_getBlockSize(int ILP, uint64_t dim_size) {
+  uint64_t block_size = 1;
+  uint64_t max_block_size = std::min(dim_size / ILP, static_cast<uint64_t>(max_threads));
+  while (block_size < max_block_size) block_size *= 2;
+  // Launch at least a single warp - the kernel assumes that.
+  block_size = std::max(block_size, static_cast<uint64_t>(32));
```
Comment (jjsjann123) [inline apex/contrib/csrc/xentropy/xentropy_kernel.cu]: Feel free to keep it as is. If we have block_size = std::max(block_size, 32) why don't we start with block_size = 32.

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+    for (int64_t i = dim + 1; i < input.dim(); ++i)
+      inner_size *= input.size(i);
+    // This kernel spawns a block per each element in the batch.
+    // XXX: it assumes that inner_size == 1
+    AT_CHECK(inner_size == 1, "Currently only inner size 1 supported");
+
+    const int ILP = 2;
+    dim3 grid(outer_size);
+    dim3 block = SoftMax_getBlockSize(ILP, dim_size);
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.type(), "host_softmax_xentropy", [&] {
```
Comment (jjsjann123) [inline apex/contrib/csrc/xentropy/xentropy_kernel.cu]: IIRC, AT_DISPATCH_xxx is deprecated(? I'm pinging @ptrblck to confirm) and we are moving towards the new dispatch: https://github.com/NVIDIA/apex/blob/master/csrc/welford.cu#L849

The new one doesn't support double yet (but could be added). But could be added if you need it.

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu (hunk 4 above)
Comment (alpha0422) [reply apex/contrib/csrc/xentropy/xentropy_kernel.cu]: Thanks for point it out! This is based on pytorch softmax code, I'll leave it here to make the diff against raw code simpler in case of we can't write the fused code base on pytorch code.

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu (hunk 5 above)
Comment (ptrblck) [reply apex/contrib/csrc/xentropy/xentropy_kernel.cu]: Yes, we moved to our dispatching (e.g. `DISPATCH_FLOAT_AND_HALF`), which seems to be already changed in the last commit.
Let me know, if we need `double` and I can add it quickly.

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+
+template<template<typename, typename, typename> class Epilogue>
+std::vector<Tensor> host_softmax_xentropy(
+        const Tensor & input_,
+        const Tensor & labels_,
+        const float smoothing,
+        const bool half_to_float){
+  if (half_to_float) AT_ASSERTM(input_.type().scalarType() == ScalarType::Half,"conversion is supported for Half type only");
+  AT_ASSERTM(labels_.type().scalarType() == ScalarType::Long,"Label type should be CUDA Long");
+  auto input = input_.contiguous();
```
Comment (ngimel) [inline apex/contrib/csrc/xentropy/xentropy_kernel.cu]: Why are you erroring out above in CHECK_INPUT, if you are making input contiguous anyway?

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+  static_assert(std::is_same<acc_type<at::Half, true>, float>::value, "accscalar_t for half should be float");
+  Tensor losses = at::empty_like(labels_, input_.options().dtype(ScalarType::Float));
+  if (input.dim() == 0) input = input.view(1);
+  int64_t dim = 1;
+  AT_ASSERTM(input.dim() == 2, "Currently only 2 dim input supported");
+  AT_CHECK(dim >=0 && dim < input.dim(), "dim must be non-negative and less than input dimensions");
+  int64_t outer_size = 1;
+  int64_t dim_size = input.size(dim);
+
+  if (input.numel() > 0) {
```
Comment (ngimel) [inline apex/contrib/csrc/xentropy/xentropy_kernel.cu]: So suppose dim_size is 0. Then uninitialized losses and max_log_sum_exp will be returned.

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+    const at::Tensor &max_log_sum_exp,
+    const at::Tensor &labels,
+    const float smoothing,
+    bool half_to_float) {
+  int64_t dim = 1;
+  Tensor gI = at::empty_like(logits_);
+  if (grad_loss.numel() == 0) {
+    return gI;
+  }
+  auto grad = grad_loss.contiguous();
```
Comment (ngimel) [inline apex/contrib/csrc/xentropy/xentropy_kernel.cu]: Same comment wrt CHECK_INPUT

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+  static_assert(std::is_same<acc_type<at::Half, true>, float>::value, "accscalar_t for half should be float");
+  if (grad.dim() == 0) grad = grad.view(1);
+  AT_CHECK(dim >=0 && dim < logits_.dim(), "dim must be non-negative and less than input dimensions");
```
Comment (ngimel) [inline apex/contrib/csrc/xentropy/xentropy_kernel.cu]: Same comment wrt dim checking

File: setup.py [hunk 10]
```diff
@@ -128,6 +128,31 @@ def check_cuda_torch_binary_vs_bare_metal(cuda_dir):
...
+            version_ge_1_1 = ['-DVERSION_GE_1_1']
+        ext_modules.append(
+            CUDAExtension(name='xentropy_cuda',
+                          sources=['apex/contrib/csrc/xentropy/interface.cpp',
+                                   'apex/contrib/csrc/xentropy/xentropy_kernel.cu'],
+                          include_dirs=['csrc'],
+                          extra_compile_args={'cxx': ['-O3'] + version_ge_1_1,
+                                              'nvcc':['-O3',
+                                                      '-gencode',
+                                                      'arch=compute_70,code=sm_70'] + version_ge_1_1}))
```
Comment (ngimel) [inline setup.py]: Why are we hardcoding arch to compute_70? So much for "generally usable"

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu [hunk 11]
```diff
@@ -0,0 +1,552 @@
...
+#include <ATen/cuda/CUDAContext.h>
+
+#include <ATen/AccumulateType.h>
+#include <ATen/cuda/NumericLimits.cuh>
+
+#include <THC/THC.h>
+#include <THC/THCGeneral.h>
+#include <THC/THCThrustAllocator.cuh>
+
+#include "type_shim.h"
```
[the remark this answers is not in the document]
Comment (mcarilli) [reply apex/contrib/csrc/xentropy/xentropy_kernel.cu]: type_shim.h is where I chose to put my dispatch macros.  It could be moved.

File: apex/contrib/csrc/xentropy/interface.cpp [hunk 12]
```diff
@@ -0,0 +1,56 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/numpy.h>
+#include <pybind11/stl.h>
```
Comment (mcarilli) [inline apex/contrib/csrc/xentropy/interface.cpp]: I don't think these lines are necessary.  <torch/extension.h> should be sufficient.

**Comment by mcarilli** [comment]
@alpha0422 Is the test included as part of the PR?  I don't see `test_label_smoothing.py ` among the contributed files.

**Comment by alpha0422** [comment]
@mcarilli No, the unittest is not in the PR. What's the proper place to put it in?

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu (hunk 11 above)
[the remark this answers is not in the document]
Comment (alpha0422) [reply apex/contrib/csrc/xentropy/xentropy_kernel.cu]: Is there any solution better than DISPATCH_FLOAT_AND_HALF? otherwise I'll leave as it is.

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu (hunk 5 above)
Comment (alpha0422) [reply apex/contrib/csrc/xentropy/xentropy_kernel.cu]: Thanks! We don't need double for now.

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu (hunk 6 above)
Comment (alpha0422) [reply apex/contrib/csrc/xentropy/xentropy_kernel.cu]: I'll leave the contiguous here, change previous CHECK_INPUT to CHECK_CUDA.

File: setup.py (hunk 10 above)
Comment (alpha0422) [reply setup.py]: Actually it's copied from previous code, removed.

File: apex/contrib/csrc/xentropy/xentropy_kernel.cu [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+        const Tensor & input_,
+        const Tensor & labels_,
+        const float smoothing,
+        const bool half_to_float){
+  if (half_to_float) AT_ASSERTM(input_.type().scalarType() == ScalarType::Half,"conversion is supported for Half type only");
+  AT_ASSERTM(labels_.type().scalarType() == ScalarType::Long,"Label type should be CUDA Long");
+  auto input = input_.contiguous();
+  Tensor max_log_sum_exp = at::empty_like(labels_, half_to_float ? input.options().dtype(ScalarType::Float) : input.options());
+  static_assert(std::is_same<acc_type<at::Half, true>, float>::value, "accscalar_t for half should be float");
+  Tensor losses = at::empty_like(labels_, input_.options().dtype(ScalarType::Float));
```
[the remark this answers is not in the document]
Comment (alpha0422) [reply apex/contrib/csrc/xentropy/xentropy_kernel.cu]: Added double into the assert.
